### ISSUE TITLE
Clojure spec split in alpha16 into spec.alpha breaks the same code in 1.8/1.9 approach

### ISSUE THREAD

**Issue by kkinnear** [tonsky/clojure-future-spec#5]
I have recently converted a library [zprint](https://github.com/kkinnear/zprint) to work with spec in both 1.8 and 1.9-alpha15, using clojure-future-spec to make 1.8 compatible (from a spec viewpoint) with 1.9.  It has worked great, and I have been delighted with how easy it was to build  a library that had the same code in both 1.9 and 1.8 and used spec.  In 1.9-alpha15 it uses the "built-in" spec, and in 1.8 people have to add clojure-future-spec as a dependency, which is a small thing.  Thank you so much for creating this library so that we can use spec in both versions!

Of course, now things have changed and the namespaces for spec have all changed in 1.9-alpha16.  This breaks my "one library works in 1.9 or 1.8" approach, since the "works in 1.8" part requires that the namespaces in clojure-future-spec be the same as the namespaces in 1.9-alpha16.

That said, now that spec is a library, it would seem that I could just add it as a dependency for my 1.8 code and use that.  But in the announcement, Alex said that spec wasn't going to support 1.8 since it uses new functions that only exist in 1.9 (and in clojure-future-spec).  I tried various approaches to adding `clojure.future` as a dependency and requiring it, but I can't even get that to work by adding it to the `(ns ...)` because the spec.alpha code looks for the new functions explicitly in clojure.core (via an alias).  I don't see a way around that, but if there is one, I'd love to hear about it.

Even if that worked (which it doesn't for me), I don't know of a way to get the clojure-future-spec `clojure.future` namespace loaded along with the new spec libraries only in 1.8 and not have it throw an error in 1.9.  There might be a way to put the require in a `(try ... catch)`, but I can't even try that because it doesn't work at all the way things are now even with the require in the `(ns ...)` declaration.

I'm hoping that at some point you will back-port the latest 1.9-alpha16 spec stuff for 1.8, as you have before with the alpha15 code.

It has been really great to have one spec that we can use in both 1.9 and 1.8.  Thank you for making that possible.  I really appreciate it, and I know I'm not the only one!

**Comment by tonsky** [comment]
Just released 1.9.0-alpha16. Should work the same (note the namespace changes, new `.alpha` suffix was added to all of them). Let me know if that works for you

**Comment by kkinnear** [comment]
Thanks for much for doing this!  Unfortunately, it doesn't actually work for me.  I have a moderately large (but not terribly complex) spec to validate a map of user options.  It works in 1.9-alpha15, and 1.8 with clojure-future-spec 1.9-alpha15.  It also works in 1.9-alpha16, but not with 1.8 and clojure-future-spec 1.9-alpha16.  Here it is in 1.9-alpha16:
```
zprint.core=> (in-ns 'zprint.spec)
#object[clojure.lang.Namespace 0x5c74f1cf "zprint.spec"]
zprint.spec=> (s/valid? ::options {:width 80})
true
```
Here it is in 1.8 with clojure-future-spec 1.9-alpha16:
```
zprint.core=> (in-ns 'zprint.spec)
#object[clojure.lang.Namespace 0x1c77fdd9 "zprint.spec"]
zprint.spec=> (s/valid? ::options {:width 80})

UnsupportedOperationException nth not supported on this type: PersistentArrayMap  clojure.lang.RT.nthFrom (RT.java:947)
zprint.spec=> (pst 100)
UnsupportedOperationException nth not supported on this type: PersistentArrayMap
	clojure.lang.RT.nthFrom (RT.java:947)
	clojure.lang.RT.nth (RT.java:897)
	clojure.spec.alpha/map-spec-impl/reify--5778 (alpha.clj:796)
	clojure.spec.alpha/conform (alpha.clj:151)
	clojure.spec.alpha/conform (alpha.clj:147)
	clojure.spec.alpha/dt (alpha.clj:749)
	clojure.spec.alpha/dt (alpha.clj:744)
	clojure.spec.alpha/dt (alpha.clj:745)
	clojure.spec.alpha/dt (alpha.clj:744)
	clojure.spec.alpha/merge-spec-impl/reify--5921/fn--5922 (alpha.clj:1157)
	clojure.core/map/fn--4789 (core.clj:2651)
	clojure.lang.LazySeq.sval (LazySeq.java:40)
	clojure.lang.LazySeq.seq (LazySeq.java:49)
	clojure.lang.RT.seq (RT.java:521)
	clojure.core/seq--4357 (core.clj:137)
	clojure.core/some (core.clj:2591)
	clojure.core/some (core.clj:2583)
	clojure.spec.alpha/merge-spec-impl/reify--5921 (alpha.clj:1158)
	clojure.spec.alpha/valid? (alpha.clj:761)
	clojure.spec.alpha/valid? (alpha.clj:757)
	zprint.spec/eval6894 (form-init3542067820488489619.clj:1)
	zprint.spec/eval6894 (form-init3542067820488489619.clj:1)
	clojure.lang.Compiler.eval (Compiler.java:6927)
	clojure.lang.Compiler.eval (Compiler.java:6890)
	clojure.core/eval (core.clj:3105)
	clojure.core/eval (core.clj:3101)
	clojure.main/repl/read-eval-print--7408/fn--7411 (main.clj:240)
	clojure.main/repl/read-eval-print--7408 (main.clj:240)
	clojure.main/repl/fn--7417 (main.clj:258)
	clojure.main/repl (main.clj:258)
	clojure.main/repl (main.clj:174)
	clojure.tools.nrepl.middleware.interruptible-eval/evaluate/fn--648 (interruptible_eval.clj:87)
	clojure.core/apply (core.clj:646)
	clojure.core/with-bindings* (core.clj:1881)
	clojure.core/with-bindings* (core.clj:1881)
	clojure.tools.nrepl.middleware.interruptible-eval/evaluate (interruptible_eval.clj:85)
	clojure.tools.nrepl.middleware.interruptible-eval/evaluate (interruptible_eval.clj:55)
	clojure.tools.nrepl.middleware.interruptible-eval/interruptible-eval/fn--693/fn--696 (interruptible_eval.clj:222)
	clojure.tools.nrepl.middleware.interruptible-eval/run-next/fn--688 (interruptible_eval.clj:190)
	java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1142)
	java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:617)
	java.lang.Thread.run (Thread.java:745)
nil
```
I haven't had a chance to try to boil this down into a small reproducible use case.  The spec I'm using is entirely in [this](https://github.com/kkinnear/zprint/blob/master/src/zprint/spec.cljc) file.  I'm using a development version, but I'm pretty sure I haven't messed with the spec in this file for a while.

So, to recap -- the spec I've been using works in:

- 1.9-alpha15
- 1.8 w/clojure-future-spec 1.9-alpha15
- 1.9-alpha16

but not in 1.8 w/clojure-future-spec 1.9-alpha16

I'll try to find something more reproducible tomorrow night.  I may have something wrong in my environment, but there isn't much there, so I don't know. I'm out of time for tonight, I'll see if I can make this fail with something easy to reproduce tomorrow night.  Thanks again for doing clojure-future-spec!

**Comment by tonsky** [comment]
Try `1.9-alpha16-1`. There was a regression we’ve introduced in `1.9-alpha16`

**Comment by kkinnear** [comment]
That fixed it, thanks!  Works great, and I have 84 spec tests where I force specific errors and see what comes out, and they all pass.  You can close this as far as I'm concerned.  Thanks again for jumping on this so quickly!
